WalletWasabi.Tests/UnitTests/WabiSabi/Backend/PhaseStepping/StepInputRegistrationTests.cs

```diff
@@ -32,6 +34,33 @@ public async Task RoundFullAsync()
 			await arena.StopAsync(CancellationToken.None);
 		}
 
+		[Fact]
+		public async Task DetectSpentTxoBeforeSteppingIntoConnectionConfirmationAsync()
+		{
+			WabiSabiConfig cfg = new() { MaxInputCountByRound = 3 };
+			var round = WabiSabiFactory.CreateRound(cfg);
+			var offendingAlice = WabiSabiFactory.CreateAlice(round); // this Alice spent the coin after registration
+
+			var mockRpc = WabiSabiFactory.CreatePreconfiguredRpcClient();
+			mockRpc.Setup(rpc => rpc.GetTxOutAsync(offendingAlice.Coin.Outpoint.Hash, (int)offendingAlice.Coin.Outpoint.N, true))
+				.ReturnsAsync((GetTxOutResponse)null!);
+
+			using Arena arena = await WabiSabiFactory.CreateAndStartArenaAsync(cfg, mockRpc, round);
+
+			round.Alices.Add(WabiSabiFactory.CreateAlice(round));
+			round.Alices.Add(offendingAlice);
+			round.Alices.Add(WabiSabiFactory.CreateAlice(round));
+			await arena.TriggerAndWaitRoundAsync(TimeSpan.FromSeconds(21));
+			Assert.Equal(Phase.InputRegistration, round.Phase);
+
+			round.Alices.Add(WabiSabiFactory.CreateAlice(round));
```

Comment:
FMI: 

* Does this line simulates something like "In InputRegistration phase, I can add Alice"? 
* Is there a risk that the Alice is not taken into account between lines L54 and L58? Then the number of Alices taken into account would not be 4 but only 3 and the test might always pass.

Reply:
> * Does this line simulates something like "In InputRegistration phase, I can add Alice"?

No, it's part of the the set up, since `round.Alices` bypasses everything and just mutates the state directly

> * Is there a risk that the Alice is not taken into account between lines L54 and L58? Then the number of Alices taken into account would not be 4 but only 3 and the test might always pass.

In general, yes, but rare - basically we don't use any form of mock time for the periodic task stuff in the tests (a problem IMO, but arguably it's too late for that). On CI tests are run in parallel so this kind of race actually seems to matter more than you'd expect, hence the fairly long timeouts.

For the specific scenario you describe, I don't think so, before line 54 there should be at most 3, but one of them is double spent and the test is confirming that that is effectively like 2 inputs, since one of them was double spent. The assertion that it is still input registration, even though when there are 3 inputs the round is considered full. That implies that `offendingAlice` was not actually registered.

When the final alice is injected, then the round can proceed to connection confirmation, because 3 unspent inputs have been registered.

Reply:
The test creates a `round` that accepts up to three alices. This means that once there are `3` alices registered the round should pass from `InputRegistration --> ConnectionConfirmation`, however this is doesn't execute concurrently at all, what we do is trigger the this verification explicitly and wait for it so, there is no risk the number of alices is not correct.

About the first question: the test registers `3` alices and triggers the round execution. in case everything is okay, the round should pass to `ConnectionConfirmation` but given it detects there is an alice that spent its registered coin it cannot pass and remains in the same `InputRegistration` phase with the only 2 valid alices. That line simply add a new valid alice to the round, something that is possible only because it is still in `InputRegistration`.

Reply:
> That line simply add a new valid alice to the round, something that is possible only because it is still in `InputRegistration`.

uh... i think "should" is missing there, we have tests that rely on injecting alices at arbitrary phases unfortunately ;-)